### ISSUE TITLE
Converter broken?

### ISSUE THREAD

**Issue by Kirin-kun** [deepfakes/faceswap#368]
With latest pull, I can't convert anymore. Did I forget something? 

```
python c:\users\kirin\faceswap\faceswap.py convert -i H:\fakes\data_B -o h:\fakes\data_B\merged -m H:\Fakes\model -S -b 4 -e 2 -D cnn
Input Directory: H:\fakes\data_B
Output Directory: h:\fakes\data_B\merged
Filter: filter.jpg
Using json serializer
Starting, this may take a while...
Loading Model from Model_Original plugin...
loaded model weights
Loading Convert from Convert_Masked plugin...
Reading alignments from: H:\fakes\data_B\alignments.json
  0%|                                                 | 0/1988 [00:00<?, ?it/s]Failed to convert image: H:\fakes\data_B\pic0001.png. Reason: expected str, bytes or os.PathLike object, not NoneType
....
etc
```

I tracked it down to this lines in convert.py:
```

if self.input_aligned_dir is not None and self.check_skipface(filename, idx):
                        print ('face {} for frame {} was deleted, skipping'.format(idx, os.path.basename(filename)))
                        continue
```

If I comment them (or specify the aligned dir with -a), it does as if it converts, but actually, it just copies the files as-is and doesn't convert anything.

What gives? 

It trains normally. I had a loss of 0.017 for A

**Comment by torzdf** [comment]
I've seen this before but can't remember the issue....

If at all possible, could you test with this PR #367 as that whole section has been rewritten, so rather than hunt down a bug in potentially obsolete code, it may be fixed in this future commit.

Alternatively you can clone from the refactor branch in my repo, which is the same code, if you would like to keep it separate.

**Comment by Kirin-kun** [comment]
I fetched your pull request, and I don't have the "expected str..." messages anymore, but Masker converter is broken: it doesn't convert anything.

I added "-c Adjust" and it then started to put new faces on the merged faces (though it's really ugly).

So, something is broken in Masked?

**Comment by Kirin-kun** [comment]
I don't understand. I re-extracted everything and tried to convert: it just copied the source images untouched, EXCEPT a false positive, where it just added some garbage.

![gamer0012](https://user-images.githubusercontent.com/36699120/39097430-58c5fd52-465c-11e8-9d82-2104bc623d57.png)

**Comment by torzdf** [comment]
Can you execute with the -v flag please.

**Comment by torzdf** [comment]
^^ I'm guessing you'll get a load of "'face {} for frame {} was deleted, skipping" errors.

If you do, please could you change (from my PR):

```
    def convert(self, converter, item):
        """ Apply the conversion transferring faces onto frames """
        try:
            filename, image, faces = item
            skip = self.opts.check_skipframe(filename)

            if not skip:
                for idx, face in faces:
                    image = self.convert_one_face(converter, (filename, image, idx, face))
            if skip != "discard":
                filename = str(self.output_dir / Path(filename).name)
                self.images.cv2_read_write('write', filename, image)
        except Exception as err:
print("Failed to convert image: {}. Reason: {}".format(filename, err))
```

to:
```
    def convert(self, converter, item):
        """ Apply the conversion transferring faces onto frames """
        filename, image, faces = item
        skip = self.opts.check_skipframe(filename)

        if not skip:
            for idx, face in faces:
                image = self.convert_one_face(converter, (filename, image, idx, face))
        if skip != "discard":
            filename = str(self.output_dir / Path(filename).name)
            self.images.cv2_read_write('write', filename, image)
```

**Comment by torzdf** [comment]
FWIW, the issue I had with this was a cuda issue, fixed with a reboot.

**Comment by Kirin-kun** [comment]
I rebooted, nothing changed. The target frames are left as-is.

BUT, for the false positive I mentioned above, with verbose, it printed:

```
Reading alignments from: H:\fakes\gamer\alignments.json
  1%|?                                       | 13/1988 [00:08<22:20,  1.47it/s]Note: Found more than one face in an image! File: H:\fakes\data_B\pic0012.png
```

And it just put garbage at the "second face" place, while leaving the "normal face" untouched.

WTF?

**Comment by Kirin-kun** [comment]
I even renamed the "aligned" directory, so it would convert everything eventually, but it did nothing.

**Comment by Kirin-kun** [comment]
It gets weirder and weirder. 

Actually, it DOES convert. but it puts reconstructed original face at almost the exact same place, so I had to flip between extracted frame and converted frame to notice it that it indeed modified the image.

I tried to swap (-s) and then it placed something else as the face, but it's something that doesn't look much like the preview, more like something not properly learned.

I'm pretty sure I trained with the good -A and -B, because the preview is normal, with AAB, AAB, BBA, BBA on each line and a loss of 0.017

**Comment by Kirin-kun** [comment]
I unearthed older trained model h5 files that I was satisfied with, using the same A face, and it converted normally. I didn't train so it's not "perfect", but the result is there.

So... is there some inversion of the layers here?

**Comment by Kirin-kun** [comment]
lool.. I trained the older model for a few minutes, it saved, and now it puts B -> B face as above.

It destroyed my good files -_-

**Comment by torzdf** [comment]
Which model are you using? There was a change implemented recently for swapping on IAE, although it looks like you are using the Original Model, so I doubt it's that.

Looking at your errors, It looks like it doesn't like moving the model to a pre-existing back up folder (on Windows at least). This is strange, and I can't see why this would corrupt your model.

I'll look at this code tomorrow, to see if something more robust can be created for backing up, which... ironically... was to protect against model corruption!

**Comment by Kirin-kun** [comment]
Yeah, I use the original model. 

But I think it swaps somewhere it shouldn't.

**Comment by torzdf** [comment]
I'm looking at the histories of 

plugins/Convert_Masked.py, 
plugins/Model_Original/
scripts/convert.py

and I can't see any changes that would effect a swap. I also did a ton of training and converting yesterday (in excess of 50k frames) and everything ran smoothly... :confused:

**Comment by Kirin-kun** [comment]
In the current master, I narrowed down the error about "expected str, bytes or os.PathLike object, not NoneType" down to the skipface function.

```
   def check_skipface(self, filename, face_idx):
        aligned_face_name = '{}_{}{}'.format(Path(filename).stem, face_idx, Path(filename).suffix)
        aligned_face_file = Path(self.arguments.input_aligned_dir) / Path(aligned_face_name)
        # TODO: Remove this temporary fix for backwards compatibility of filenames
        bk_compat_aligned_face_name = '{}{}{}'.format(Path(filename).stem, face_idx, Path(filename).suffix)
        bk_compat_aligned_face_file = Path(self.arguments.input_aligned_dir) / Path(bk_compat_aligned_face_name)
        return aligned_face_file not in self.input_aligned_dir and bk_compat_aligned_face_file not in self.input_aligned_dir
```

The problem is obvious I think? The line:

`aligned_face_file = Path(self.arguments.input_aligned_dir) / Path(aligned_face_name)`

uses the "self.arguments.input_aligned_dir", which is None if no -a option has been given. BOOM.

**Comment by Kirin-kun** [comment]
Still have the problem with my conversion being swapped though. It's driving me mad.

**Comment by torzdf** [comment]
That first issue **should** be fixed in my refactor that I hope to push to staging today.

I'm still working on the assumption that it is the backup function that is somehow corrupting your model, so I'm going to work on a fix.

It would be really useful to me if you could pull the last commit prior to the backup being implemented and test in the meantime (so pull this commit: 9dc151e5b58abb5f8862d2aa84124ed86156e0b8)

**Comment by Kirin-kun** [comment]
That's exactly what I'm doing. Great men think alike :smile: 

I just reverted to 9dc151e

I reused the same model files and suddenly, instead of 0.017, the loss shot up to 0.03x, even though the preview almost looks like how it looked at 0.017

I see hope!

**Comment by torzdf** [comment]
Ok, I have raised a PR just for the Original model at the moment #371 

Please could you test, and if it fixes the issue I will implement for all models and push.

**Comment by Kirin-kun** [comment]
I tested conversion both with your PR and with 9dc151e and it still insists on putting face B to B.

It's maddening. I'm sure my arguments are good, because I just made two shell scripts, one for training and one for conversion and I feed them the exact same arguments: A faceset name and B faceset name. 

And the preview looks ok.

To train:
`python c:\users\kirin\faceswap\faceswap.py train -A H:\Fakes\%1\aligned -B H:\Fakes\%2\aligned -m H:\Fakes\modelo.%1.%2 -p -bs 64 -ag
`

To convert:
`python c:\users\kirin\faceswap\faceswap.py convert -i H:\fakes\%2 -o h:\fakes\%2\merged -m H:\Fakes\modelo.%1.%2 -b 4 -e 2 -S -D cnn -a h:\fakes\%2\aligned`

But then, at convert time, it reconstructs B face on output frame instead of A face.

I'm stumped.

I'm going to try with completely different face sets and start from scratch.

**Comment by Kirin-kun** [comment]
FWIW:
pip freeze

absl-py==0.2.0
astor==0.6.2
bleach==1.5.0
click==6.7
cycler==0.10.0
decorator==4.3.0
dlib==19.10.0
face-recognition==1.2.2
face-recognition-models==0.3.0
gast==0.2.0
grpcio==1.11.0
h5py==2.7.1
html5lib==0.9999999
Keras==2.1.5
keras-contrib==2.0.8
keras-vggface==0.5
kiwisolver==1.0.1
Markdown==2.6.11
matplotlib==2.2.2
networkx==2.1
numpy==1.14.2
opencv-python==3.3.0.10
pathlib==1.0.1
Pillow==5.1.0
protobuf==3.5.2.post1
pyparsing==2.2.0
python-dateutil==2.7.2
pytz==2018.4
PyWavelets==0.5.2
PyYAML==3.12
scandir==1.7
scikit-image==0.13.1
scipy==1.0.1
six==1.11.0
tensorboard==1.7.0
tensorflow-gpu==1.7.0
termcolor==1.1.0
tqdm==4.23.0
Werkzeug==0.14.1

**Comment by torzdf** [comment]
Ok, this is frustrating!

Seeing as no one else seems to be having this issue, then I can only assume it must be at your end.

I will probably still push my backup fix, as it resolves the issue with losing any files stored in your model dir.

Could you confirm that you now have .bk files inside your model_dir after running from my PR?

**Comment by Kirin-kun** [comment]
Yes, I have .bk files in the model dir.

I just made a small training session with random new facesets and checked convert: it puts A face on B when I specify "-s" to swap.

Train window looks like this, which is OK in my book:

![train1](https://user-images.githubusercontent.com/36699120/39122199-6ef1e6ca-46f4-11e8-9a1d-c9ed6bb170fc.png)

But with the conversion (without the -s), it gives this:

![face1](https://user-images.githubusercontent.com/36699120/39122351-eb4cee9a-46f4-11e8-94d3-3f25bb944ea3.png)

Which is obviously B->B

So, without -s, I have B -> B and with -s I have A -> B

Just, WHY?

**Comment by torzdf** [comment]
Wait, what? Your source frames are wrong, aren't they?

That source frame looks like it's from B...

> To train:
> python c:\users\kirin\faceswap\faceswap.py train -A H:\Fakes\%1\aligned -B H:\Fakes\%2\aligned -m H:\Fakes\modelo.%1.%2 -p -bs 64 -ag
> 
> To convert:
> python c:\users\kirin\faceswap\faceswap.py convert -i H:\fakes\%2 -o h:\fakes\%2\merged -m H:\Fakes\modelo.%1.%2 -b 4 -e 2 -S -D cnn -a h:\fakes\%2\aligned

The convert should be:
`python c:\users\kirin\faceswap\faceswap.py convert -i H:\fakes\%1 -o h:\fakes\%2\merged -m H:\Fakes\modelo.%1.%2 -b 4 -e 2 -S -D cnn -a h:\fakes\%2\aligned`

**Comment by torzdf** [comment]
So it should be:

`train -A <Original Face> -B <Replacement Face>`

`convert -i <Original Frames, where A came from> -m <A to B model>`

**Comment by Kirin-kun** [comment]
FCK.

I've been inverting my arguments. I checked an old cmd and indeed I swapped the A and B for the training. I don't know what came over me.

I need a new brain. 

Thanks for your time (well, at least I found a bug in the skipface thx to this, so it's not *totally* wasted).

**Comment by torzdf** [comment]
Lol. It's all good. If I'd paid closer attention, I would have realised that it was right there in your very first post ;)

> python c:\users\kirin\faceswap\faceswap.py convert -i **H:\fakes\data_B** -o h:\fakes\data_B\merged -m H:\Fakes\model -S -b 4 -e 2 -D cnn
> Input Directory: H:\fakes\data_B
